bigflow/workflow.py

```diff
@@ -29,8 +30,36 @@ def daily_start_time(start_time: dt.datetime) -> dt.datetime:
     return start_time.replace(hour=0, minute=0, second=0, microsecond=0) - td
 
 
+class JobContext(typing.NamedTuple):
+    runtime_raw: typing.Optional[str]
+    runtime: typing.Optional[dt.datetime]
+    env: typing.Optional[str]
+    workflow: typing.Optional['Workflow']
+    workflow_id: typing.Optional[str]
+    # TODO: add unique 'workflow execution id' (for tracing/logging)
+    # TODO: add some 'workflow-level' configuration here
+
+
+class Job(abc.ABC):
+    """Base abstract class for bigflow.Jobs.  It is recommended to inherit all your jobs from this class."""
+
+    @property
+    @abc.abstractmethod
+    def id(self) -> str:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def execute(self, context: JobContext):
+        raise NotImplementedError
+
+
 class Workflow(object):
 
+    RUNTIME_FORMATS = [
```

Comment:
Why do we need to support multiple runtime formats? Currently, for a deployed workflow, a user can't choose the format of a passed runtime.

Reply:
AFAIK in most places we use '%y-%m-%d' as runtime format.
But also when runtime is not specified - it is autogenerated to be in '%y-%m-%d %H:%M:%S' format.

Reply:
Users can run a workflow with runtime formatted in any way. But when scheduled, the only available format is date-time:

dagbuilder.py
```
 arguments=['run', '--job', '{bf_job}', '--runtime', '{{{{ execution_date.strftime("%Y-%m-%d %H:%M:%S") }}}}', '--project-package', '{root_folder}', '--config', '{{{{var.value.env}}}}'],
```

Handling different formats is currently strictly related to the CLI run. And as I can see, CLI already has that logic:

```
def _valid_datetime(dt):
    try:
        datetime.strptime(dt, "%Y-%m-%d %H:%M:%S")
        return dt
    except ValueError:
        try:
            datetime.strptime(dt, "%Y-%m-%d")
            return dt
        except ValueError:
            raise ValueError("Not a valid date: '{0}'.".format(dt))
```

Reply:
I saw '%y-%m-%d' in unit tests. https://github.com/allegro/bigflow/blob/master/test/test_workflow.py#L17
https://github.com/allegro/bigflow/blob/master/test/test_scaffold_bigquery_example.py#L23
I thought it is an API, do we need to update tests?

As I can see _validate_datetime only performs validation, it doesn't change actual 'dt' string.

Reply:
"As I can see _validate_datetime only performs validation, it doesn't change actual 'dt' string." - true. I think it can be useful for daily workflows (so a user doesn't have to pass time every time running `bf run`).

"I saw '%y-%m-%d' in unit tests. https://github.com/allegro/bigflow/blob/master/test/test_workflow.py#L17
https://github.com/allegro/bigflow/blob/master/test/test_scaffold_bigquery_example.py#L23
I thought it is an API, do we need to update tests?" - There should be a test for each format in this case.